In [ ]:
2import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

In [ ]:
df = pd.read_csv('iphone_purchase_records.csv')
df


,Gender,Age,Salary,Purchase Iphone
0,Male,19,19000,0
1,Male,35,20000,0
2,Female,26,43000,0
3,Female,27,57000,0
4,Male,19,76000,0
...,...,...,...,...
395,Female,46,41000,1
396,Male,51,23000,1
397,Female,50,20000,1
398,Male,36,33000,0


In [ ]:
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()

In [ ]:
label.fit(df['Gender'])

LabelEncoder()

In [ ]:
df['Gender_Encoded'] = label.transform(df['Gender'])

In [ ]:
df = df.drop('Gender', axis=1)

In [ ]:
df.head()

,Age,Salary,Purchase Iphone,Gender_Encoded
0,19,19000,0,1
1,35,20000,0,1
2,26,43000,0,0
3,27,57000,0,0
4,19,76000,0,1


In [ ]:
X = df.drop('Purchase Iphone', axis=1)
Y = df['Purchase Iphone']

In [ ]:
X_train, X_test, Y_train, Y_test=train_test_split(X,Y, test_size=0.3, random_state=42)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
k_values = range(1,21)
accuracy_scores = []

for k in k_values:
  knn = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
  knn.fit(X_train, Y_train)
  Y_pred = knn.predict(X_test)
  accuracy = accuracy_score(Y_test, Y_pred)
  accuracy_scores.append(accuracy)

optimal_k = k_values[accuracy_scores.index(max(accuracy_scores))]
print('Optimal k:', optimal_k)

Optimal k: 5


In [ ]:
knn=KNeighborsClassifier(n_neighbors=5)

In [ ]:
knn.fit(X_train, Y_train)

KNeighborsClassifier()

In [ ]:
y_pred=knn.predict(X_test)

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score

In [ ]:
precision = precision_score(Y_test, y_pred, average='weighted')
print(f"Precision: {precision}")

Precision: 0.8266631835597354


In [ ]:
recall = recall_score(Y_test, y_pred, average='weighted')
print(f"Recall: {recall}")

Recall: 0.8166666666666667


In [ ]:
f1_score = f1_score(Y_test, y_pred, average='weighted')
print(f"f1: {f1_score}")

f1: 0.8086458333333333


In [ ]:
input_data =  {}

for feature in X.columns:
  while True:
    try:
      if feature.lower() in ['gender', 'gender_encoded']:
        value = input(f"Enter value for {feature} (Male = 1, Female = 0): ")
        if value not in ['1', '0']:
          print("Invalid input. Please enter 1 for Male or 0 for Female")
          continue
        value = int(value)
      else:
        value = float(input(f'Enter value for {feature}: '))

      input_data[feature] = value
      break
    except ValueError:
      print("Invalid input. Please enter a numerical value.")


input_df= pd.DataFrame([input_data])

prediction = knn.predict(input_df)[0]

if prediction == 1:
  if input_data['Salary'] > 70000:
    print("Prediction: Likely to purchase IPhone. Target as a premium customer.")
  elif 10000 < input_data['Salary'] < 70000:
    print("Prediction: Will most likely finance IPhone")
else:
  print("Will not purchase Iphone")

Enter value for Age: 21
Enter value for Salary: 5000
Enter value for Gender_Encoded (Male = 1, Female = 0): 1
Will not purchase Iphone


Hello boss,
For this project, I implement a K-Nearest Neighbor (KNN) classification model to predict wheteher a customer will purchase an IPhone based on age, salary, and gender. I conducted feature encoding for categorical data, which was using a label encoder to switch from a strong to an integer on the gender in order to improve the model. I standardized the dataset, and optimized the model by tuing the hyperparameter k, selecting the optimal value based on accuracy. The final model achieved a precision of 0.83, recall of 0.82, and an F1 score of 0.81. Additionally, I incorporated a conditional statement in the prediction output to differentiate potential premium customers based on salary, proving a more targeted business insight.